### DISCUSSION TITLE
Removing events

### DISCUSSION THREAD

**Discussion by gtrias** [RailsEventStore/rails_event_store#1136] [General]
Hi, 
We're started using you amazing library to publish and store several events but I just realized there are some of them which shall be removed. I'm trying to figure how I can create a migration to do it but I didn't find any `destroy`, `remove` or `delete` method which can help me doing that.

Is there a way to remove already persisted events?

**Comment by mpraglowski** [comment]
Hi
Why would you like to remove events? The events are the facts that has already happened - you shpuld not change the past. 

Maybe some strategies for event versioning could help you. Please take a look at great book by Greg Young (avaiable for free to read online): https://leanpub.com/esversioning/read_full

**Comment by mpraglowski** [comment]
> Is there a way to remove already persisted events?
 
There is no API for that on RES.

**Comment by gtrias** [comment]
Thank you for the book reference :)
We accidentally persisted some non wanted events and I want to just clean event store of such events... We're in a phase of experimentation with events right now.

**Comment by mostlyobvious** [comment]
Provided you're on ActiveRecord adapter, following can help you in phase of experimentation:

```
event = RailsEventStoreActiveRecord::Event.find(...)
RailsEventStoreActiveRecord::EventInStream.where(event_id: event.id).delete_all
event.destroy
```

**Reply by prasannatamma** [reply]
RailsEventStoreActiveRecord::Event not working

**Reply by mostlyobvious** [reply]
`RailsEventStoreActiveRecord::Event` has been marked as private along the way, not to become depended on. In a special case like unsupported deletes, one can define:

```ruby
TemporaryEvent = Class.new(ActiveRecord::Base) { self.table_name = "event_store_events" }
TemporaryEventInStream = Class.new(ActiveRecord::Base) { self.table_name = "event_store_events_in_streams" }

```

**Comment by gtrias** [comment]
Thanks @pawelpacana I'm going to play with the ActiveRecord adapter right now. I'll share any success or failure here :)

**Comment by paneq** [comment]
> RailsEventStoreActiveRecord::EventInStream.where(event_id: event.id).delete_all

this is dangerous for explicitly ordered streams (that use `expected_version` - `https://railseventstore.org/docs/publish/#expected_version`) for example those created by `AggregateRoot`. It should be ok, if that's not the case.

**Comment by gtrias** [comment]
Finally I've made a simple script running the queries directly against mysql.

```
module EventSystem
  module Tooling
    class EventDestroyer
      # @param event_ids [Array<String>]
      def initialize(
        event_ids:
      )
        @event_ids = event_ids
      end

      def call
        event_ids_str = @event_ids.to_s.tr('[]','')
        ActiveRecord::Base
          .connection
          .execute("DELETE FROM event_store_events WHERE id IN(#{event_ids_str})")

        ActiveRecord::Base
          .connection
          .execute("DELETE FROM event_store_events_in_streams WHERE event_id IN(#{event_ids_str})")
      end
    end
  end
end
```
Note that `event_id` field declared in `event_store_events_in_streams` table is not a FK to `event_store_events` so, If you only remove the events the RailsEventStore will break with errors of type:

```
NoMethodError:   undefined method `id' for nil:NilClass
```

**Comment by mpraglowski** [comment]
The script should work, however https://github.com/RailsEventStore/rails_event_store/blob/master/rails_event_store_active_record/lib/rails_event_store_active_record/generators/templates/create_event_store_events_template.rb#L9 - `position` values for stream in `event_store_events_in_streams` might be invalid. This might cause some issues later.

**Comment by gtrias** [comment]
Thanks for the reply @mpraglowski ,

We have the position field set at NULL in all events, it's only filled when you use the `expected_version` functionality?

**Comment by mpraglowski** [comment]
It is explained in https://railseventstore.org/docs/expected_version/

You probably do not use `:auto` strategy, and always keep the RES default `:any`.
That's why you do not have position value set. In your case it is :+1: :)
